lib/srv/desktop/audit.go

```diff
@@ -145,6 +149,208 @@ func (s *WindowsService) onClipboardReceive(ctx context.Context, emitter events.
 	s.emit(ctx, emitter, event)
 }
 
+// onSharedDirectoryAnnounce adds the shared directory's name to the auditCache.
+func (s *WindowsService) onSharedDirectoryAnnounce(sid string, m tdp.SharedDirectoryAnnounce) {
+	s.auditCache.SetName(sessionID(sid), directoryID(m.DirectoryID), directoryName(m.Name))
+}
+
+// onSharedDirectoryAcknowledge emits a DesktopSharedDirectoryStart on a successful receipt of a
+// successful tdp.SharedDirectoryAcknowledge.
+func (s *WindowsService) onSharedDirectoryAcknowledge(
+	ctx context.Context,
+	emitter events.Emitter,
+	id *tlsca.Identity,
+	sid string,
+	desktopAddr string,
+	m tdp.SharedDirectoryAcknowledge,
+) {
+	if m.ErrCode != tdp.ErrCodeNil {
+		return
+	}
+
+	code := libevents.DesktopSharedDirectoryStartCode
+	name, ok := s.auditCache.GetName(sessionID(sid), directoryID(m.DirectoryID))
+	if !ok {
+		code = libevents.DesktopSharedDirectoryStartFailureCode
+		name = "unknown"
+		s.cfg.Log.Warnf("failed to find a directory name corresponding to sessionID(%v), directoryID(%v)", sid, m.DirectoryID)
+	}
+
+	event := &events.DesktopSharedDirectoryStart{
+		Metadata: events.Metadata{
+			Type:        libevents.DesktopSharedDirectoryStartEvent,
+			Code:        code,
+			ClusterName: s.clusterName,
+			Time:        s.cfg.Clock.Now().UTC(),
+		},
+		UserMetadata: id.GetUserMetadata(),
+		SessionMetadata: events.SessionMetadata{
+			SessionID: sid,
+			WithMFA:   id.MFAVerified,
+		},
+		ConnectionMetadata: events.ConnectionMetadata{
+			LocalAddr:  id.ClientIP,
+			RemoteAddr: desktopAddr,
+			Protocol:   libevents.EventProtocolTDP,
+		},
+		DesktopAddr:   desktopAddr,
+		DirectoryName: string(name),
+		DirectoryID:   m.DirectoryID,
+	}
+
+	s.emit(ctx, emitter, event)
+}
+
+// onSharedDirectoryReadRequest adds ReadRequestInfo to the auditCache.
+func (s *WindowsService) onSharedDirectoryReadRequest(sid string, m tdp.SharedDirectoryReadRequest) {
+	s.auditCache.SetReadRequestInfo(sessionID(sid), completionID(m.CompletionID), readRequestInfo{
```

Comment:
If the server never sends a response, will this be cached forever? Should we check after 5-10 seconds and purge it from the cache if we didn't get a response?

Reply:
No, it won't end up cached forever. At this point we create an entry like
```
s.auditCache.m[<sessionID>].readRequestCache[<completionID>] = readRequestInfo{<info>}
```

And then in `onSessionEnd` we just delete all of `s.auditCache.m[<sessionID>]`
https://github.com/gravitational/teleport/blob/c81307b2290ccc8e2a68072ff71ed0575f93f571/lib/srv/desktop/audit.go#L70-L72

That makes me realize, though, that that's the only cache cleanup we have. Which means that the `readRequestCache` of a single session can technically balloon in size if it does a bunch of reads in a row (`completionID` is 32-bit unsigned). That said, in most of the use cases the `completionID`'s stay below ~10, since once an operation is completed, its `completionID` gets recycled. So cleaning up the `readRequestCache` every time a request is completed may create a significant performance hit for the most common use case, since it would require more allocs and deallocs.

My thinking is: 
1. We can comfortably get away with things as written, assuming normal circumstances
2. If we go with a timeout, it needs to be longer than 5-10 sec, since long (e.g. 1MB) reads can sometimes take a while. Probably want 45s to be safe
3. If we really want to harden this to potential accidental or malicious memory DOS attack, we'll need to think of something more clever

Reply:
It will only be cached until the end of the session

Reply:
48dadacb4524515277d9deb03e9a1c28041b6841

Reply:
I think the timer is unnecessary. Requests without corresponding replies should be very uncommon, if it's happening to any significant degree then there's something wrong with the session and we can expect its soon to be ended (at which point the entire cache corresponding to that sessionID will be cleared).

Reply:
f879e5e184d8a817e8c9a15f053eafbde3bb45ae

Reply:
If this is the case and we need to add a read timeout, then I think you'd have to do that with your own `select` inside a function in go, since the tdpConn wrapper keeps the connection as a bufio Reader which can't set a deadline

Reply:
We use this activity monitor:

https://github.com/gravitational/teleport/blob/dc581ac08300310f66ea1191115d83d0ed34b5fa/lib/srv/desktop/windows_server.go#L861-L891

which will end the connection if it doesn't receive any messages from the client for too long (the line below is called in the loop that receives incoming TDP messages):

https://github.com/gravitational/teleport/blob/dc581ac08300310f66ea1191115d83d0ed34b5fa/lib/srv/desktop/rdp/rdpclient/client.go#L310

Reply:
That said, this system is vulnerable to a memory DOS attack -- i.e. if a user modifies the client to drop all shared directory read requests, and then convinces RDP to send a bunch in a row (it wouldn't be difficult to write a program that does that by just reading from the shared file repeatedly), then the `readRequestCache` for that session would explode in size.

The only way I see to account for that is to limit the overall size of the `readRequestCache` (i.e. only allow max 1000 entries at a time, which I think should be more than enough given that I've never seen it go higher than 10 during development). A timed cleanup couldn't work, because the timeout would need to be too long (say 10 sec) to actually prevent a programmatic attack, which could send a billion read requests by then.

What do you think @zmb3, something we should account for?

Reply:
Implemented this in several steps:

6ff00671cda0317ae615bf506fb1524b8fa126d5
22f449eb63b7b854955aab39533660b0c89b982e
74008eb961da0d6ae5db71a19cda9df0dece09f0
87a64a2388c5a1352af807ed7a3a47a24441848d